Dev/mech micro

marcorusc opened the pull request:
Closes #26, #28 , #29 , #37 , #45 , #54

marcorusc committed:
feat(mechanics): add simulation parameters struct

- Introduce simulation_parameters struct to centralize mechanics configuration
- Define interaction_config for potential-specific settings
- Add support for type-based interaction parameters
- Resolves #26

marcorusc committed:
feat(mechanics): define mechanical agent data

- Add agent_data struct for Structure-of-Arrays (SoA) storage
- Implement agent proxy class for object-oriented access
- Add unit tests for agent data management and accessors
- Resolves #28

marcorusc committed:
feat(mechanics): implement agent container

- Add agent_container class for managing agent lifecycle
- Implement create() and remove() operations
- Ensure proper synchronization with base_agent_data
- Add unit tests for container operations
- Resolves #29

marcorusc committed:
feat(mechanics): implement serialization of mechanical state

- Add vtk_mechanics_serializer for VTK output
- Implement serialization of agent positions, velocities, and forces
- Add cell_data structure for aggregated cell-level data
- Resolves #37

marcorusc committed:
feat(mechanics): add solver interface and registry

- Define abstract solver base class
- Implement solver_registry for runtime backend selection
- Add unit tests for registry mechanics

marcorusc committed:
feat(mechanics): implement uniform grid spatial index

- Add abstract spatial_index interface
- Implement uniform_grid_spatial_index for efficient neighbor queries
- Add unit tests for spatial indexing

marcorusc committed:
feat(mechanics): implement interaction potentials

- Add potential_interface for force models
- Implement Standard (PhysiCell), Morse, and Kelvin-Voigt potentials
- Add unit tests for all potential types

marcorusc committed:
feat(mechanics): implement OpenMP solver backend

- Implement OpenMP-parallelized solver class
- Add sub-solvers for forces, motility, neighbors, and positions
- Integrate potentials and spatial index into solver loop
- Add solver registration logic
- Resolves #54

marcorusc committed:
docs(mechanics): add mechanics subsystem documentation

- Add Mechanics section to README
- Describe environment, solver, and serialization usage
- Provide minimal code example
- Resolves #45

marcorusc committed:
refactor(common): move vtk_serializer_base to common library

- Move VTK serializer base class from biofvm to common
- Update biofvm serializers to use common base
- Enable code reuse for mechanics serializer

marcorusc committed:
feat(mechanics): integrate environment with solver and agents

- Update environment to use agent_container and solver_registry
- Wire up simulation loop in run_single_timestep
- Update build configuration to link new components
- Update environment tests

marcorusc committed:
automatic clang formatting fix

marcorusc committed:
adhere to the clang-tidy linter

asmelko commented on mechanics/micromechanics/include/micromechanics/agent_generic_storage.h:
```diff
@@ -0,0 +1,68 @@
+
+#include "agent_data.h"
+#include "cell_interface.h"
+
```
add pragma to disable windows warnings at the beginning and end of file here same as is in `reactions-diffusion/biofvm/include/biofvm/agent_generic_storage.h`

asmelko commented on mechanics/micromechanics/include/micromechanics/cell_data.h:
```diff
@@ -0,0 +1,135 @@
+	// ========== Methods ==========
+
+	/// Clear all cell data.
+	void clear()
```
the implementations of the cell methods should go to the src

asmelko commented on mechanics/micromechanics/include/micromechanics/environment.h:
```diff
@@ -1,20 +1,71 @@
 
-	void run_single_timestep() override;
+	/// Cell-level data (pressure, etc.) aggregated from agents
+	cell_data cells;
```
I see that environment is holding just cell_data but not agent_data

asmelko replied:
cells should be in a cell_container, similar as agents are

asmelko commented on mechanics/micromechanics/include/micromechanics/simulation_parameters.h:
```diff
@@ -0,0 +1,159 @@
+	 *
+	 * Adds config for both (type_a, type_b) and (type_b, type_a).
+	 */
+	void add_interaction(std::uint8_t type_a, std::uint8_t type_b, const interaction_config& config)
```
this should be in src

asmelko reviewed (changes requested):
Part 1

asmelko commented on mechanics/micromechanics/kernels/openmp_solver/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+target_link_libraries(
+  physicore.mechanics.micromechanics.kernels.openmp_solver_internal_iface
+  INTERFACE physicore.mechanics.micromechanics_internal_iface
+            physicore.mechanics.micromechanics_iface physicore::common)
```
internal_iface is ment to be internal, so only for testing.

asmelko commented on mechanics/micromechanics/tests/test_potentials.cpp:
```diff
@@ -0,0 +1,332 @@
+using namespace physicore::mechanics::micromechanics;
+using namespace physicore::mechanics::micromechanics::kernels::openmp_solver;
+
+class PotentialTest : public ::testing::Test
```
since potentials implementation is in the kernel, this test should be in the kernel's tests. also, I would appreciate the _exact_ test, not just GT and LT. do some simple examples, compute the forces on the paper by hand and then compare your implementation matches expected output

asmelko commented on mechanics/micromechanics/tests/test_agent.cpp:
```diff
@@ -0,0 +1,150 @@
+using namespace physicore;
+using namespace physicore::mechanics::micromechanics;
+
+class AgentTest : public ::testing::Test
```
but i think its fine here - but remember to update this when you change agent.h

asmelko commented on mechanics/micromechanics/tests/test_agent_data.cpp:
```diff
@@ -0,0 +1,165 @@
+using namespace physicore;
+using namespace physicore::mechanics::micromechanics;
+
+class AgentDataTest : public ::testing::Test
```
test_agent_data should verify that all arrays in agent_data have correct size after add and remove, see reactions-diffusion/biofvm/tests/test_agent_data.cpp. just have here 2 tests in the same fashion as in biofvm

asmelko commented on mechanics/micromechanics/include/micromechanics/spatial_index.h:
```diff
@@ -0,0 +1,20 @@
+
+class environment;
+
+class spatial_index
```
maybe spacial index should also go to the kernel implementation?

marcorusc replied:
For now it stays shared - it's used by environment for any kernel. Moving to per-kernel makes sense long-term (GPU kernels may use different structures? I have no idea)

joseestragues replied:
It think it can stay shared by the moment, GPU kernels may use this same implementation to search neighbours ids and build the data to be transfered to device (or activate pairs execution). This is how I visualize a GPU kernel using this same methods.

asmelko replied:
but building of the index is also a costly method suitable for parallelization. GPU-impl will be different from CPU. Lets go on and put this into OpenMP kernel and ideally also parallelize it

asmelko commented on mechanics/micromechanics/include/micromechanics/environment.h:
```diff
@@ -1,20 +1,71 @@
+	simulation_parameters params;
+
+	/// The active solver (obtained from solver_registry)
+	std::unique_ptr<solver> active_solver;
```
maybe just solver is fine. calling it active_solver suggests that there is a time when solver is inactive which is not the case

marcorusc replied:
Fixed in 1e666e3. Renamed active_solver → solver.

asmelko reviewed (changes requested):
I suggest dividing this PR into 3:
- serialization
- openmp_solver
- the rest that will remain this PR

marcorusc commented on mechanics/micromechanics/include/micromechanics/agent_data.h:
```diff
@@ -0,0 +1,257 @@
+	physicore::base_agent_data_generic_storage<ContainerType>& base_data;
+
+	// Agent Classification
+	ContainerType<std::uint8_t> agent_types; // Agent type for type-based interactions (0-255)
```
yes we will never pass the 256 agent types... maximum I saw was 10, if we add the compartments we can arrive at 30 maybe but it is a huge stretch...

marcorusc committed:
fix(micromechanics): silence MSVC dominance warnings

marcorusc committed:
fix(openmp_solver.cpp): fix standalone-agent checks

marcorusc committed:
refactor (micromechanics): rename active_solver to solver

marcorusc committed:
second automatic clang formatting fix

marcorusc committed:
fix(micromechanics/cell_data.h): moved header implementations to src

marcorusc committed:
refactor(micromechanics): move simulation_parameters impl to src

marcorusc committed:
refactor(micromechanics): decouple environment from agent_container include

marcorusc committed:
chore(docs): remove README changes from PR

marcorusc committed:
refactor(micromechanics): make VTK mechanics serialization non-selective

marcorusc committed:
refactor(micromechanics): move deps off internal iface

marcorusc committed:
refactor(openmp_solver): keep internal iface tests-only

marcorusc committed:
refactor(micromechanics): move VTK mechanics serialization out

marcorusc committed:
refactor(micromechanics): move OpenMP kernel out

marcorusc committed:
refactor(micromechanics): move solver_registry to common

Implements micromechanics solver registration via a header-only common factory registry and drops the old solver_registry_sole.cpp implementation.

marcorusc committed:
refactor(micromechanics): replace cell_data maps with SoA

Converts cell-level storage to flat SoA vectors and adds cell_generic_storage to provide cell-level accessors, updating tests accordingly.

marcorusc committed:
fix(micromechanics): remove stored cell speed

Drops leftover speed references; speed is derived from velocity on demand. Includes a minimal OpenMP backend adjustment to avoid writing removed fields.

marcorusc committed:
refactor(micromechanics): initialize domain bounds in ctor

Moves environment domain_min/domain_max defaults from the header into the environment constructor.

marcorusc committed:
refactor(micromechanics): hide compartment agent interface

marcorusc committed:
refactor(micromechanics): simplify to 1-cell-1-compartment SEM model

Simplify the micromechanics module to a minimal working framework where
each cell is represented by exactly one agent (1-cell-1-compartment).

Data model changes:
- Replace agent_data unordered_maps with flat SoA vectors
- Simplify cell_data to flat vectors indexed by cell_id with cell_offset()
- Remove cell_definition layer and compartment_offset logic
- Add cell_aggregation for computing cell-level properties from agents
- Thin out cell_interface to essential properties only
- Rename active_solver to solver in environment

Potential API changes:
- Change potential_interface::calculate_pairwise_force to return real_t
  instead of void with output parameter (addresses asmelko review)
- Update Morse potential to use r²/r₀² exponent convention
- Remove standard_potential (merged into Morse/Kelvin-Voigt)
- Fix force sign convention: potentials return signed values,
  force_solver accumulates with += for clarity

OpenMP kernel changes:
- Move std::random_device out of parallel region in motility_solver
- Restructure motility to operate at cell level then propagate to agents
- Simplify openmp_solver to delegate to cell_aggregation
- Update all solvers for flat SoA data access

Test changes:
- Update all existing tests for the simplified data model
- Add cell_aggregation test suite
- Add Morse and Kelvin-Voigt potential unit tests
- Set up OpenMP kernel test infrastructure with CMakeLists

marcorusc commented on mechanics/micromechanics/include/micromechanics/cell_data.h:
```diff
@@ -0,0 +1,135 @@
+	std::unordered_map<index_t, std::array<real_t, 3>> velocities;
+
+	/// Cell speed (velocity magnitude) indexed by cell_id.
+	std::unordered_map<index_t, real_t> speeds;
```
Removed in commit 1e666e3.  cell_data no longer stores  speeds. Velocity vectors are sufficient and speed can be derived when needed.

marcorusc commented on mechanics/micromechanics/include/micromechanics/cell_interface.h:
```diff
@@ -0,0 +1,40 @@
+	virtual std::uint8_t& is_movable() = 0;
+
+	// Per-agent interaction strengths (can vary within a cell type)
+	virtual real_t& cell_cell_adhesion_strength() = 0;
```
Addressed in commit 1e666e3. In the 1-cell-1-compartment implementation (simplified), forces act on agents directly. Potential parameters live in simulation_parameters (per type-pair interaction_config). cell_interface is now a thin wrapper for agent-level properties.

marcorusc commented on mechanics/micromechanics/include/micromechanics/environment.h:
```diff
@@ -1,20 +1,71 @@
+	std::array<real_t, 3> domain_min = { -500.0, -500.0, -500.0 };
+
+	/// Domain boundaries [x_max, y_max, z_max]
+	std::array<real_t, 3> domain_max = { 500.0, 500.0, 500.0 };
```
Fixed in 296ad7b (domain bounds init in ctor) and 1e666e3 (renamed active_solver → solver). Domain defaults [-500, 500]³ match PhysiCell for compatibility.

marcorusc replied:
But of course they can be changed

marcorusc commented on mechanics/micromechanics/include/micromechanics/potential_interface.h:
```diff
@@ -0,0 +1,51 @@
+	 * @param dz Z-component of normalized direction vector
+	 * @param force_out Output: force magnitude (positive = repulsion, negative = attraction)
+	 */
+	virtual void calculate_pairwise_force(const environment& env, index_t agent_i, index_t agent_j, real_t distance,
```
Fixed in 1e666e3. calculate_pairwise_force now returns  real_t (positive=repulsion, negative=adhesion). Force solver accumulates with +=, eliminating the double-negation.

marcorusc commented on mechanics/micromechanics/include/micromechanics/cell_data.h:
```diff
@@ -0,0 +1,135 @@
+ * All properties are indexed by cell_id. Properties are computed
+ * by aggregating data from agents belonging to each cell.
+ */
+struct cell_data
```
Fixed in commit 1e666e3.  cell_data no longer uses unordered_map. All properties are now flat SoA vectors indexed by  cell_id, with a cell_offset(cell_id, dim) helper for 3D data

marcorusc commented on mechanics/micromechanics/include/micromechanics/uniform_grid_spatial_index.h:
```diff
@@ -0,0 +1,38 @@
+	}
+};
+
+class uniform_grid_spatial_index : public spatial_index
```
Same as above, maybe revisit when adding a second kernel backend?

marcorusc committed:
clang-format correction

marcorusc committed:
fix(openmp_solver): remove static from stateful sub-solver methods

Remove the static keyword from update_neighbors, update_motility,
update_interactions, update_spring_attachments, and update_positions.
These methods are invoked through instance members that maintain
internal state (e.g. initialized_ flag), so they should be non-static
member functions. Fixes readability-static-accessed-through-instance
clang-tidy warnings in openmp_solver.cpp.

marcorusc committed:
fix: make update_neighbors and update_positions static to resolve clang-tidy warnings

marcorusc committed:
fix: call static methods via class name to satisfy readability-static-accessed-through-instance

marcorusc committed:
fix: resolve SonarQube issues in uniform_grid_spatial_index.cpp

- Make mech_data_ptr a reference-to-const
- Extract collect_neighbors_in_cell to reduce nesting depth and cognitive complexity

marcorusc committed:
fix: resolve clang-tidy warnings for static methods, dead stores, and global variables

- Make update_motility, update_interactions, update_spring_attachments static
- Update call sites in openmp_solver.cpp to use class-name qualification
- Remove dead stores (dt, attachment_rate, detachment_rate) in spring_solver.cpp
- Move registered to anonymous namespace as const in register_solver.cpp

marcorusc committed:
fix: resolve clang-tidy warnings in environment, cell_aggregation, uniform_grid

- Initialize domain_min/domain_max in member initializer list (cppcoreguidelines-pro-type-member-init)
- Use std::ranges::fill instead of std::fill (modernize-use-ranges)
- Replace redundant int type with auto where static_cast makes type obvious

marcorusc committed:
fix: make vtk_serializer_base member variables private instead of protected

- Move iteration, output_dir, vtks_dir, pvd_file_name, pvd_contents to private
- Add protected accessors: iteration(), advance_iteration(), vtks_dir()
- Update subclasses vtk_serializer and vtk_agents_serializer to use accessors

marcorusc committed:
fix: add const to test variables and use anonymous namespace for test helper

- Add const to morse_potential, kelvin_voigt_potential, and force variables in test_potentials.cpp
- Move add_agent helper from static to anonymous namespace in test_cell_aggregation.cpp

marcorusc committed:
fix: resolve clang-tidy warnings in test_agent_container and test_cell_data

- Move make_agent_container to anonymous namespace (misc-use-anonymous-namespace)
- Add NOLINTNEXTLINE for cognitive complexity in parameterized test (gtest macro expansion)
- Add const to cell c0 in test_cell_data (misc-const-correctness)

marcorusc committed:
fix: use in-class initializers for domain_min/domain_max instead of constructor init list

Satisfies SonarQube rule: use in-class initializer instead of constructor's initializer list

marcorusc committed:
fix: reduce nesting depth in query_neighbors by using early continue

Replaces if(found){process} with if(not found){continue} to stay within
SonarQube's 3-level nesting limit for if/for/do/while/switch.

marcorusc committed:
test: add coverage for simulation_parameters and environment

- New test_simulation_parameters.cpp: covers add_interaction (symmetric + self-pair),
  get_interaction (found + default fallback), set_single_type_interaction,
  default values for both simulation_parameters and interaction_config
- Expanded test_environment.cpp: constructor defaults, initialize_solver via registry,
  run_single_timestep with mock solver testing all feature flag combinations
  (motility, basement membrane, spring attachments), serialize_state

joseestragues commented on common/include/common/factory_registry.h:
```diff
@@ -0,0 +1,100 @@
+ * link dependencies.
+ */
+template <typename ProductT>
+class factory_registry
```
I also moved factory_register to common. Structure is very similar. We can address it in the future.

joseestragues commented on mechanics/micromechanics/include/micromechanics/agent_data.h:
```diff
@@ -0,0 +1,101 @@
+	physicore::base_agent_data_generic_storage<ContainerType>& base_data;
+
+	// Agent Classification
+	ContainerType<std::uint8_t> compartment_types; // Agent type for type-based interactions (0-255)
```
Adam suggested me to use vector instead of ContainerType for PhysiCell::agent_data. Is this the same case for sem? shall both be equal?

marcorusc replied:
They should, I will adapt it

joseestragues commented on mechanics/micromechanics/include/micromechanics/cell_aggregation.h:
```diff
@@ -0,0 +1,17 @@
+namespace physicore::mechanics::micromechanics {
+
+/// Reset per-cell aggregate buffers that are computed from agents.
+void reset_cell_aggregates(cell_data& cells);
```
Would be worth specific aggregation methods for each metric? In order to reduce compute cost

marcorusc replied:
Good idea, I will add this

asmelko commented:
can you actually do the PR division that I mentioned? I still see the serialization and openmp code present here

asmelko commented on common/include/common/vtk_serializer_base.h:
```diff
@@ -8,24 +8,29 @@
-	std::string vtks_dir;
-	std::string pvd_file_name;
+	[[nodiscard]] std::size_t iteration() const { return iteration_; }
+	void advance_iteration() { ++iteration_; }
 
-	std::string pvd_contents;
+	[[nodiscard]] const std::string& vtks_dir() const { return vtks_dir_; }
```
move implementation to cpp

asmelko commented on mechanics/micromechanics/include/micromechanics/agent_data.h:
```diff
@@ -0,0 +1,101 @@
+	// Physics State
+	ContainerType<real_t> velocities;		   // n * dims
+	ContainerType<real_t> previous_velocities; // n * dims (for Adams-Bashforth)
+	ContainerType<real_t> forces; // n * dims magnitude of forces acting on agent (we can keep it for later)
```
what is forces once again? if we are not using it now, lets remove it

asmelko commented on mechanics/micromechanics/include/micromechanics/agent_data.h:
```diff
@@ -0,0 +1,101 @@
+	ContainerType<real_t> forces; // n * dims magnitude of forces acting on agent (we can keep it for later)
+
+
+	// Topology (Kelvin-Voigt)
+	// Note: Nested vectors are not ideal for SoA but required for dynamic topology
```
remove the comments

asmelko commented on mechanics/micromechanics/include/micromechanics/agent_generic_storage.h:
```diff
@@ -0,0 +1,103 @@
+
+namespace internal {
+
+/**
+ * @brief Internal interface for mechanics agents (sub-cellular compartments).
+ *
+ * This is intentionally not part of the public micromechanics API surface.
+ */
+class agent_interface : public virtual base_agent_interface
+{
+public:
+	// Agent Classification
+	virtual std::uint8_t& compartment_type() = 0;
+	virtual index_t& cell_id() = 0;
+
+	// Kinematics
+	virtual std::span<real_t> velocity() = 0;
+	virtual std::span<real_t> previous_velocity() = 0;
+	virtual std::span<real_t> force() = 0;
+
+	// Topology (Kelvin-Voigt)
+	virtual std::span<index_t> spring_attachments() = 0;
+};
```
this should be in a separate file

asmelko commented on mechanics/micromechanics/include/micromechanics/cell_data.h:
```diff
@@ -0,0 +1,144 @@
+template <template <typename...> typename ContainerType = std::vector>
+struct cell_data_generic_storage
+{
+	static constexpr index_t invalid_cell_id = static_cast<index_t>(-1);
```
i dont see when we would need this

asmelko commented on mechanics/micromechanics/include/micromechanics/cell_data.h:
```diff
@@ -0,0 +1,144 @@
+	index_t cells_count = 0;
+
+	// ========== Per-cell physical properties ==========
+	ContainerType<real_t> radii;			// cells_count — shared agent radius
```
so call it agent_radius

asmelko commented on mechanics/micromechanics/include/micromechanics/cell_data.h:
```diff
@@ -0,0 +1,144 @@
+	// ========== Helpers ==========
+
+	/// Index into a per-cell vector field (positions, velocities, …).
+	std::size_t cell_offset(index_t cell_id, index_t dim) const
+	{
+		return static_cast<std::size_t>(cell_id) * static_cast<std::size_t>(dims) + static_cast<std::size_t>(dim);
+	}
```
this method should not be a part of this struct's contract. You can put helpers into the tests, where you are using it

asmelko commented on mechanics/micromechanics/include/micromechanics/cell_data.h:
```diff
@@ -0,0 +1,144 @@
+	void clear();
+
+	/// Resize / grow storage for `new_cells_count` cells in `new_dims` dimensions.
+	void resize(index_t new_cells_count, index_t new_dims);
```
remove this method - cell_data should have a constructor, which accepts dims

asmelko commented on mechanics/micromechanics/include/micromechanics/cell_data.h:
```diff
@@ -0,0 +1,144 @@
+	}
+
+	/// Clear all cell data vectors (count stays unchanged — call resize to reset).
+	void clear();
+
+	/// Resize / grow storage for `new_cells_count` cells in `new_dims` dimensions.
+	void resize(index_t new_cells_count, index_t new_dims);
+
+	/// Convenience overload — keeps current dims (must have been set previously).
+	void resize(index_t new_cells_count);
```
please, let this struct have the same interface as other data structs -> add() and remove_at(index)

asmelko commented on mechanics/micromechanics/include/micromechanics/environment.h:
```diff
@@ -1,20 +1,68 @@
+	simulation_parameters params;
+
+	/// The solver (obtained from solver_registry)
+	std::unique_ptr<solver> solver_;
```
why underscore only in this member?

asmelko commented on mechanics/micromechanics/include/micromechanics/environment.h:
```diff
@@ -1,20 +1,68 @@
+	void run_single_timestep() override;
 	void serialize_state(real_t current_time) override;
+
+	// Note: definitions live here because they are global simulation configuration
+	// shared across solver backends.
```
remove comment

asmelko commented on mechanics/micromechanics/include/micromechanics/environment.h:
```diff
@@ -1,20 +1,68 @@
+	explicit environment(real_t timestep);
+	~environment() override;
+
+	/**
+	 * @brief Initialize the solver from the registry.
+	 *
+	 * Must be called after setting params.solver_name.
+	 */
```
i think this sould be inside of the method, ideally just asserted without the comment

asmelko commented on mechanics/micromechanics/include/micromechanics/cell_aggregation.h:
```diff
@@ -0,0 +1,17 @@
+
+namespace physicore::mechanics::micromechanics {
+
+/// Reset per-cell aggregate buffers that are computed from agents.
+void reset_cell_aggregates(cell_data& cells);
+
+/// Aggregate COM position/velocity, per-compartment counts, and a pressure proxy from agent forces.
+void aggregate_cell_data_from_agents(const physicore::base_agent_data& base, const agent_data& agents,
+									 cell_data& cells);
+
+} // namespace physicore::mechanics::micromechanics
```
this should be backend specific

asmelko commented on mechanics/micromechanics/tests/test_agent_container.cpp:
```diff
@@ -0,0 +1,155 @@
+class RemoveAgentTest : public ::testing::TestWithParam<std::tuple<int, bool>>
+{};
+
+// NOLINTNEXTLINE(readability-function-cognitive-complexity)
```
instead of these comments, copy .clang-format from biofvm's test file into here

marcorusc commented:
Hi Adam, I am resuming work on micromechanics and have reviewed the updated branches (lots of updates!) and your latest comment.
I propose rebuilding #79 as a core-only PR containing the cell/sub-agent data model, containers, parameters, environment, and abstract solver interfaces. OpenMP, including spatial-index construction and cell aggregation, would remain in #106, while VTK serialization would remain in #105.
Before rewriting the branch, can you confirm:
- should feature/mech-sem still be the integration target and be updated from current main
- should I update #79 in place after archiving its current head, or open a replacement PR
- should micromechanics use the factory_registry implementation now present on feature/mech-orig
- should moving the VTK serializer base to common be handled as a separate PR associated with #104

I will keep the current branches archived and address the existing review threads explicitly in the rebuilt PR. Is that ok?
